digest/src/core_api.rs

```diff
@@ -117,3 +117,15 @@ pub enum TruncSide {
     /// Truncate right side, i.e. `&out[m..]`.
     Right,
 }
+
+/// Core trait for saving internal state of the hash core and restoring it later.
+pub trait HashCoreInternalState {
+    /// Internal state of the hash core.
+    type InternalState;
```

Comment:
It'd be good to have some bounds here to enable generic code:

```suggestion
    type InternalState: AsRef<[u8]>;
```

Reply:
If the internal state consists of several objects, is it possible to implement `TryFrom<&[u8]> + AsRef<[u8]>`: e.g. internal state of SHA-2 hashes consists of the intermediate hash value (array of `u32`/`u64`) and the number of consumed blocks (number `u64`/`u128`).

Reply:
The use case of this API is serialization.

If you think there's a use case beyond serialization, can you explain how that use case isn't satisfied by simply cloning an existing digest instance?

Reply:
Yes, the use case is serialization only.
Should I then create a byte array and serialize the internal state on my own?
E.g. for `Sha256VarCore`, I need to allocate an array, copy the intermediate hash value (`[u32; STATE_LEN]`), and the number of consumed blocks (`u64`).
Or is it better to return an object that consists of the intermediate hash value and the number of consumed blocks, and this object is `Serializable` (`serde`)? But this object can not implement `AsRef<[u8]>` because it consists of a byte array and number.

Reply:
It depends if the state can always be fixed-width or not. If it can, we should probably use an associated `generic_array::ArrayLength<u8>` instead.

I think it's a pretty reasonable assumption that most digests could support a fixed-width serializable state.